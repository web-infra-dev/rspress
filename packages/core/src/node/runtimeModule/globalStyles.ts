import fs from '@rspress/shared/fs-extra';
import { type FactoryContext, RuntimeModuleID } from '.';

export async function globalStylesVMPlugin(context: FactoryContext) {
  const { config, pluginDriver } = context;
  const globalStylesByPlugins = pluginDriver.globalStyles();

  if (config.globalStyles) {
    // Only patch user config global styles
    const source = config.globalStyles;
    const styleContent = await fs.readFile(source, 'utf-8');

    // Patch --modern-abc, .modern-abc global name
    const patchedStyleContent = styleContent
      .replace(/--modern-/g, '--rp-')
      .replace(/\.modern-doc/g, '.rspress')
      .replace(/\.modern-/g, '.rspress-');

    await fs.writeFile(source, patchedStyleContent, 'utf-8');
  }

  const moduleContent = (
    await Promise.all(
      [config?.globalStyles || '', ...globalStylesByPlugins]
        .filter(source => source.length > 0)
        .map(async source => {
          return `import ${JSON.stringify(source)};`;
        }),
    )
  ).join('');

  return {
    [RuntimeModuleID.GlobalStyles]: moduleContent,
  };
}
