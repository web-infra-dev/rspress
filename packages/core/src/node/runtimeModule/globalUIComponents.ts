import { type FactoryContext, RuntimeModuleID } from '.';

export async function globalUIComponentsVMPlugin(context: FactoryContext) {
  const { config, pluginDriver } = context;
  let index = 0;

  const globalUIComponentsByPlugins = pluginDriver.globalUIComponents();
  const globalComponents = [
    ...(config?.globalUIComponents || []),
    ...globalUIComponentsByPlugins,
  ];
  const moduleContent = globalComponents
    .map(source => {
      const name = `Comp_${index}`;
      if (Array.isArray(source)) {
        return `import ${name} from ${JSON.stringify(source[0])};
const Props_${index++} = ${JSON.stringify(source[1])};\n`;
      }

      index++;
      return `import ${name} from ${JSON.stringify(source)};\n`;
    })
    .concat(
      `export default [${Array.from({ length: index }, (_, i) => {
        if (Array.isArray(globalComponents[i])) {
          return `[${`Comp_${i}`}, Props_${i}]`;
        }
        return `Comp_${i}`;
      }).join(', ')}];`,
    )
    .join('');

  return {
    [RuntimeModuleID.GlobalComponents]: moduleContent,
  };
}
