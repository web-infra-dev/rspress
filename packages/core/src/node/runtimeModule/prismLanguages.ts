import { DEFAULT_HIGHLIGHT_LANGUAGES } from '@rspress/shared';
import { uniqBy } from 'lodash-es';
import { FactoryContext, RuntimeModuleID } from '.';

function getLanguageMeta(highlightLanguages: (string | string[])[]) {
  const aliases: Record<string, string[]> = {};
  const languageMeta = uniqBy(
    [...DEFAULT_HIGHLIGHT_LANGUAGES, ...highlightLanguages].map(language => {
      const isArray = Array.isArray(language);
      const [alias, name] = isArray ? language : [language, language];

      if (isArray) {
        const temp = aliases[name] || (aliases[name] = []);

        if (!temp.includes(alias)) {
          temp.push(alias);
        }
      }

      return { alias, name };
    }),
    'name',
  );
  return { languageMeta, aliases };
}

export async function prismLanguageVMPlugin(context: FactoryContext) {
  const { config } = context;
  const { highlightLanguages = [] } = config.markdown || {};

  const { languageMeta, aliases } = getLanguageMeta(highlightLanguages);

  const importStatement = languageMeta.map(language => {
    const { alias, name } = language;
    return `import ${alias} from 'react-syntax-highlighter/dist/esm/languages/prism/${name}'`;
  });

  const moduleContent = `
${importStatement.join('\n')}

export const aliases = ${JSON.stringify(aliases)};

export default {
  ${languageMeta.map(({ alias, name }) => {
    return `"${name}": ${alias}`;
  })}
};
`;
  return {
    [RuntimeModuleID.PrismLanguages]: moduleContent,
  };
}
