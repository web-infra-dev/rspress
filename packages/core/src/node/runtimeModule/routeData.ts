import { FactoryContext, RuntimeModuleID } from '.';

export async function routeVMPlugin(context: FactoryContext) {
  const { routeService, isSSR } = context;

  // client: The components of route is lazy loaded
  if (!isSSR) {
    return {
      [RuntimeModuleID.RouteForClient]: routeService.generateRoutesCode(false),
    };
  }

  return {
    [RuntimeModuleID.RouteForClient]: routeService.generateRoutesCode(false),
    [RuntimeModuleID.RouteForSSR]: routeService.generateRoutesCode(true),
  };
}
