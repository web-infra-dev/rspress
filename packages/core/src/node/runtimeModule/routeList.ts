import { RuntimeModuleID, type VirtualModulePlugin } from './types';

// TODO: hmr with addContextDependency
export const routeListVMPlugin: VirtualModulePlugin = context => {
  const { routeService } = context;

  return {
    [RuntimeModuleID.Routes]: () => {
      return routeService.generateRoutesCode();
    },
  };
};
