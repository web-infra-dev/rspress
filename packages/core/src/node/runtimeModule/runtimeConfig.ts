import { normalizeSlash } from '@rspress/shared';
import { RuntimeModuleID, type VirtualModulePlugin } from './types';

/**
 * some configuration in `rspress.config.ts` serialized from compile time side to runtime
 */
export const runtimeConfigVMPlugin: VirtualModulePlugin = context => {
  return {
    [RuntimeModuleID.RuntimeConfig]: () => {
      const { config } = context;

      const { base } = config;

      // TODO: base can be normalized in compile time side in an earlier stage
      const normalizedBase = normalizeSlash(base ?? '/');
      // Use named export for better tree-shaking support
      return `export const base = ${JSON.stringify(normalizedBase)};
export const ssg = ${JSON.stringify(config.ssg ?? true)};`;
    },
  };
};
