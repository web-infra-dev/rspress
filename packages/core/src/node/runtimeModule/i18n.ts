import { join } from 'node:path';
import type { TransformHandler } from '@rsbuild/core';
import type { UserConfig } from '@rspress/shared';
import { logger } from '@rspress/shared/logger';
import { type FactoryContext, RuntimeModuleID } from '.';

const DEFAULT_I18N_SOURCE = join(process.cwd(), 'i18n.json');

export function getI18nData(docConfig: UserConfig) {
  const { i18nSourcePath = DEFAULT_I18N_SOURCE } = docConfig;
  try {
    // require.cache is an API in Rslib.
    delete REQUIRE_CACHE[i18nSourcePath];
    // eslint-disable-next-line import/no-dynamic-require
    const i18nSource = require(i18nSourcePath);
    return i18nSource;
  } catch (e) {
    logger.debug('getI18nData Failed: \n', e);
    return {};
  }
}

/**
 * Generate i18n text for client runtime
 */
export function i18nVMPlugin(
  context: Omit<FactoryContext, 'isSSR' | 'alias'>,
): Record<string, TransformHandler> {
  const { config } = context;
  return {
    [RuntimeModuleID.I18nText]: ({ addDependency, addMissingDependency }) => {
      addDependency(config.i18nSourcePath || DEFAULT_I18N_SOURCE);
      addMissingDependency(config.i18nSourcePath || DEFAULT_I18N_SOURCE);
      const i18nData = getI18nData(config);

      return `export default ${JSON.stringify(i18nData, null, 2)}`;
    },
  };
}
