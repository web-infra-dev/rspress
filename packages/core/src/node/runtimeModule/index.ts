import type { UserConfig } from '@rspress/shared';
import { RspackVirtualModulePlugin } from 'rspack-plugin-virtual-module';
import { RsbuildPlugin } from '@rsbuild/core';
import { RouteService } from '../route/RouteService';
import { PluginDriver } from '../PluginDriver';
import { routeVMPlugin } from './routeData';
import { siteDataVMPlugin } from './siteData/index';
import { i18nVMPlugin } from './i18n';
import { globalUIComponentsVMPlugin } from './globalUIComponents';
import { globalStylesVMPlugin } from './globalStyles';
import { searchHookVMPlugin } from './searchHooks';

export interface FactoryContext {
  userDocRoot: string;
  config: UserConfig;
  isSSR: boolean;
  runtimeTempDir: string;
  alias: Record<string, string | string[]>;
  routeService: RouteService;
  pluginDriver: PluginDriver;
}

type RuntimeModuleFactory = (
  context: FactoryContext,
) => Record<string, string> | Promise<Record<string, string>>;

export const runtimeModuleFactory: RuntimeModuleFactory[] = [
  /**
   * Generate route data for client and server runtime
   */
  routeVMPlugin,
  /**
   * Generate search index and site data for client runtime
   * Also responsible for automatically importing prism languages
   */
  siteDataVMPlugin,
  /**
   * Generate global components from config and plugins
   */
  globalUIComponentsVMPlugin,
  /**
   * Generate global styles from config and plugins
   */
  globalStylesVMPlugin,
  /**
   * Generate i18n text for client runtime
   */
  i18nVMPlugin,
  /**
   * Generate search hook module
   */
  searchHookVMPlugin,
];

// We will use this plugin to generate runtime module in browser, which is important to ensure the client have access to some compile-time data
export function rsbuildPluginDocVM(
  factoryContext: Omit<FactoryContext, 'alias'>,
): RsbuildPlugin {
  const { pluginDriver } = factoryContext;
  return {
    name: 'rsbuild-plugin-doc-vm',
    setup(api) {
      api.modifyBundlerChain(async bundlerChain => {
        // The order should be sync
        const alias = bundlerChain.resolve.alias.entries();
        const runtimeModule: Record<string, string> = {};
        // Add internal runtime module
        for (const factory of runtimeModuleFactory) {
          const moduleResult = await factory({
            ...factoryContext,
            alias: alias as Record<string, string>,
          });
          Object.assign(runtimeModule, moduleResult);
        }
        // Add runtime module from outer plugins
        const modulesByPlugin = await pluginDriver.addRuntimeModules();
        Object.keys(modulesByPlugin).forEach(key => {
          if (runtimeModule[key]) {
            throw new Error(
              `The runtime module ${key} is duplicated, please check your plugin`,
            );
          }
          runtimeModule[key] = modulesByPlugin[key];
        });
        bundlerChain.plugin('rspress-runtime-module').use(
          // @ts-expect-error
          new RspackVirtualModulePlugin(
            runtimeModule,
            factoryContext.runtimeTempDir,
          ),
        );
      });
    },
  };
}

export enum RuntimeModuleID {
  GlobalStyles = 'virtual-global-styles',
  GlobalComponents = 'virtual-global-components',
  RouteForClient = 'virtual-routes',
  RouteForSSR = 'virtual-routes-ssr',
  SiteData = 'virtual-site-data',
  SearchIndexHash = 'virtual-search-index-hash',
  I18nText = 'virtual-i18n-text',
  SearchHooks = 'virtual-search-hooks',
  PrismLanguages = 'virtual-prism-languages',
}

export const runtimeModuleIDs = [
  RuntimeModuleID.GlobalStyles,
  RuntimeModuleID.GlobalComponents,
  RuntimeModuleID.RouteForClient,
  RuntimeModuleID.RouteForSSR,
  RuntimeModuleID.SiteData,
  RuntimeModuleID.SearchIndexHash,
  RuntimeModuleID.I18nText,
  RuntimeModuleID.SearchHooks,
];
