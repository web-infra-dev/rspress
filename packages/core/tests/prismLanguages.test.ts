import { describe, expect, it } from 'vitest';
import { prismLanguageVMPlugin } from '../src/node/runtimeModule/prismLanguages';

describe('getHighlightLanguages', () => {
  it('prismLanguageVMPlugin should be configurable to users', async () => {
    const meta = await prismLanguageVMPlugin({
      config: {
        markdown: {
          highlightLanguages: [
            // 1. add lang
            'zig',
            // 2. add alias
            ['MD', 'markdown'],
            // 3. add lang and alias
            ['oc', 'objectivec'],
          ],
        },
      },
      ...({} as any),
    });
    expect(meta).toMatchSnapshot();
  });
});
