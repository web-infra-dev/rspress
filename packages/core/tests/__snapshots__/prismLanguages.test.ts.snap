// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getHighlightLanguages > prismLanguageVMPlugin should be configurable to users 1`] = `
{
  "virtual-prism-languages": "
import js from 'react-syntax-highlighter/dist/esm/languages/prism/javascript'
import ts from 'react-syntax-highlighter/dist/esm/languages/prism/typescript'
import jsx from 'react-syntax-highlighter/dist/esm/languages/prism/tsx'
import json from 'react-syntax-highlighter/dist/esm/languages/prism/json'
import css from 'react-syntax-highlighter/dist/esm/languages/prism/css'
import scss from 'react-syntax-highlighter/dist/esm/languages/prism/scss'
import less from 'react-syntax-highlighter/dist/esm/languages/prism/less'
import xml from 'react-syntax-highlighter/dist/esm/languages/prism/xml-doc'
import diff from 'react-syntax-highlighter/dist/esm/languages/prism/diff'
import yaml from 'react-syntax-highlighter/dist/esm/languages/prism/yaml'
import md from 'react-syntax-highlighter/dist/esm/languages/prism/markdown'
import bash from 'react-syntax-highlighter/dist/esm/languages/prism/bash'
import zig from 'react-syntax-highlighter/dist/esm/languages/prism/zig'
import oc from 'react-syntax-highlighter/dist/esm/languages/prism/objectivec'

export const aliases = {\\"javascript\\":[\\"js\\"],\\"typescript\\":[\\"ts\\"],\\"tsx\\":[\\"jsx\\",\\"mdx\\"],\\"xml-doc\\":[\\"xml\\"],\\"markdown\\":[\\"md\\",\\"MD\\"],\\"objectivec\\":[\\"oc\\"]};

export default {
  \\"javascript\\": js,\\"typescript\\": ts,\\"tsx\\": jsx,\\"json\\": json,\\"css\\": css,\\"scss\\": scss,\\"less\\": less,\\"xml-doc\\": xml,\\"diff\\": diff,\\"yaml\\": yaml,\\"markdown\\": md,\\"bash\\": bash,\\"zig\\": zig,\\"objectivec\\": oc
};
",
}
`;
